operator/controllers/kyma_controller.go

```diff
@@ -121,283 +122,282 @@ func (r *KymaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
-	if err := r.ReconcileKymaForRelease(ctx, kyma, templates); err != nil {
+	// these are the actual modules
+	modules, err := util.ParseTemplates(kyma, templates)
+	if err != nil {
 		return err
 	}
 
-	// update status conditions
-	updateRequired, err := r.checkAndUpdateComponentConditions(ctx, kyma, templates)
+	statusUpdateRequiredFromCreation, err := r.CreateOrUpdateModules(ctx, kyma, modules)
 	if err != nil {
+		message := fmt.Sprintf("Component CR creation error: %s", err.Error())
+		logger.Info(message)
+		r.Event(kyma, "Warning", "ReconciliationFailed", fmt.Sprintf("Reconciliation failed: %s", message))
+		statusErr := status.Helper(r).UpdateStatus(ctx, kyma, operatorv1alpha1.KymaStateError, message)
+		if statusErr != nil {
+			return statusErr
+		}
 		return err
 	}
-	if updateRequired {
-		return r.updateKymaStatus(ctx, kyma, kyma.Status.State, "updating component conditions")
+
+	// Now we track the conditions: update the status based on their state
+	statusUpdateRequiredFromSync, err := r.SyncConditionsWithModuleStates(ctx, kyma, modules)
+	if err != nil {
+		return err
 	}
 
 	// set ready condition if applicable
 	if kyma.AreAllReadyConditionsSetForKyma() {
 		message := fmt.Sprintf("reconciliation of %s finished!", kyma.Name)
 		logger.Info(message)
 		r.Event(kyma, "Normal", "ReconciliationSuccess", message)
-		return r.updateKymaStatus(ctx, kyma, operatorv1alpha1.KymaStateReady, message)
+		return status.Helper(r).UpdateStatus(ctx, kyma, operatorv1alpha1.KymaStateReady, message)
+	}
+
+	// if the ready condition is not applicable, but we changed the conditions, we still need to issue an update
+	if statusUpdateRequiredFromCreation || statusUpdateRequiredFromSync {
+		return status.Helper(r).UpdateStatus(ctx, kyma, kyma.Status.State, "updating component conditions")
 	}
 
 	return nil
 }
 
-func (r *KymaReconciler) HandleDeletingState(ctx context.Context, logger *logr.Logger, kyma *operatorv1alpha1.Kyma) (bool, error) {
+func (r *KymaReconciler) HandleDeletingState(ctx context.Context, kyma *operatorv1alpha1.Kyma) (bool, error) {
+	logger := log.FromContext(ctx)
 	templates, err := release.GetTemplates(ctx, r, kyma)
 	if err != nil {
-		return false, fmt.Errorf("deletion cannot proceed - templates could not be fetched: %w", err)
+		return false, status.Helper(r).UpdateStatus(ctx, kyma, operatorv1alpha1.KymaStateError, "templates could not be fetched")
 	}
 
-	for _, component := range kyma.Spec.Components {
-		actualComponentStruct, err := util.GetUnstructuredComponentFromTemplate(templates, component.Name, kyma)
-		if err != nil {
-			return false, err
-		}
-
-		if err = r.Get(ctx, client.ObjectKeyFromObject(actualComponentStruct), actualComponentStruct); err == nil {
+	modules, err := util.ParseTemplates(kyma, templates)
+	if err != nil {
+		return false, err
+	}
+	for name, module := range modules {
+		if err = r.Get(ctx, client.ObjectKeyFromObject(module.Unstructured), module.Unstructured); err == nil {
```

Comment:
If we use `r.Get(ctx, client.ObjectKeyFromObject(module.Unstructured), module.Unstructured);`
does it mean we still expect component CR exists in control-plane cluster? I thought we moved it to target cluster or?

Reply:
Component CR exists on both clusters, the target cluster is only synchronized